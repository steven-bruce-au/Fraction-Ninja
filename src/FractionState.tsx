
/** Type used to by the FractionComparator to describe the state of a fraction column
*/
export class FractionState {
  numerator: number
  denominator: number
  charts: number // number of whole charts required to render this fraction 
  value: number 

  constructor(numerator: number=1, denominator: number=1) {
  	this.numerator = numerator;
  	this.denominator = denominator;
    this.value = numerator / denominator;
    this.charts = Math.ceil(numerator / denominator);
  }

}