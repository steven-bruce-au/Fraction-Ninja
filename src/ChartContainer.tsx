import React from 'react';
import memoize from "fast-memoize";

import PieFraction from './PieFraction';
import { FractionState } from './FractionState';
import styles from './ChartContainer.module.scss'

type ChartData = {
  equal: boolean,
  fractionState: FractionState,
  maxCharts: number
}

/** Chart container renders multiple PieFractions determined by the ChartData */
export default class ChartContainer extends React.Component<ChartData, {}> {

  constructor(props: ChartData) {
    super(props);
    this.state = {
    };
  }

  chartWidth = memoize(
    (charts: number, maxCharts: number)  =>  { return { width: `${charts * 100/maxCharts}%`}; }
  );

  render() : JSX.Element {
    const { numerator, denominator, charts } = this.props.fractionState;
    const { equal, maxCharts } = this.props;

    // Build an array of whole pie charts, empty if value of fraction is not greater than one
    const wholePies = new Array<JSX.Element>();
      for(let i = 1; i < charts; i++) {
        // add "whole" charts as necessary for fractions > 1
        wholePies.push(<PieFraction equal={equal} numerator={denominator} denominator={denominator}/>);
      }

     // The last pie chart, the fractional part
     const remainder = (numerator === 0) ? <PieFraction equal={equal} numerator={0} denominator={denominator}/> : 
                                         <PieFraction equal={equal} numerator={numerator - (charts - 1) * denominator} denominator={denominator}/>
    return (
       <div className={styles.chartContainer} style={this.chartWidth(charts, maxCharts)}>
            {wholePies}
            {remainder}
       </div>
    );
  }
}
