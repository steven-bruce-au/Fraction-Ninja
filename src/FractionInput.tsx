import React from 'react';
import { FractionState } from './FractionState';
import styles from './FractionInput.module.scss'

/** Type used by the FractionInput props */
export interface FractionInputProps {
  fractionState: FractionState;
    numeratorUpdated(numerator: number): void;
    denominatorUpdated(denominator: number): void;
}


/** Render the value of the fraction numerator and denominator along with the buttons to increment and decrement them
*/
export default class FractionInput extends React.Component<FractionInputProps, {}> {

  constructor(props: FractionInputProps) {
    super(props);
    this.state = {
    };

    this.handleNumeratorChange = this.handleNumeratorChange.bind(this);
    this.handleDenominatorChange = this.handleDenominatorChange.bind(this);

  }

  /** Handler for the denominator value changing, used by both increment and decrement
  @param the new denominator value
  @return a callback for the button component to trigger the change
  */
  handleDenominatorChange(denominator: number) : (event: any) => void {
    return (event: any) => {
      this.props.denominatorUpdated(denominator);
    }
  }

 /** Handler for the numerator value changing, used by both increment and decrement
  @param the new numerator value
  @return a callback for the button component to trigger the change
  */
  handleNumeratorChange(numerator: number) : (event: any) => void {
    return (event: any) => {
       this.props.numeratorUpdated(numerator);
    }
  }

  render() : JSX.Element {
    const { numerator, denominator} = this.props.fractionState;

    return (
       <div className={styles.fraction}>
         <div className={styles.fractionRow} >
          <button className={styles.button} onClick={this.handleNumeratorChange(numerator - 1)}> - </button>
          <div className={`${styles.number} ${styles.underline} ${styles.numerator}`}> { numerator }</div>
          <button className={styles.button} onClick={this.handleNumeratorChange(numerator + 1)}> + </button>
        </div>

         <div className={styles.fractionRow} >
          <button className={styles.button} onClick={this.handleDenominatorChange(denominator - 1)}>-</button>
          <div className={`${styles.number} ${styles.denominator}`}> {denominator}</div>
          <button className={styles.button} onClick={this.handleDenominatorChange(denominator + 1)}>+</button>
        </div>
       </div>
    );
  }
}
