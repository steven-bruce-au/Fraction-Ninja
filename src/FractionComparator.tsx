import React from 'react';
import { FractionState } from './FractionState';
import FractionInput from './FractionInput';
import ChartContainer from './ChartContainer';
import styles from './FractionComparator.module.scss';

enum FractionColumn {
    LEFT = 0,
    RIGHT = 1
}


type FractionComparatorState = {
  fractions: Array<FractionState>,
  equal: boolean
}

/** State is lifted up to this component. It receives events from child components and orchestrates them */
export default class FractionComparator extends React.Component<{}, FractionComparatorState> {

  constructor(props: any) {
    super(props);
    this.state = {
      fractions: [ new FractionState(), new FractionState() ],
      equal: true
    };

    this.handleNumeratorChange = this.handleNumeratorChange.bind(this);
    this.handleDenominatorChange = this.handleDenominatorChange.bind(this);
  }

  /** Construct the FractionComparatorState object based on an event modifying the current state 
    @param existingFunctions The existing state of the object
    @param numerator The new numerator for the changed fraction
    @param denominator The new denominator for the changed fraction
    @param updatedFractionColumn The fraction that has changed
    @return The new comparator state
  */
  private buildState(existingFractions: Array<FractionState>, numerator: number, denominator: number, updatedFractionColumn: FractionColumn) : FractionComparatorState {
     // copy and update the fractions
      const fractionsCopy = Array.from(existingFractions);
      fractionsCopy[updatedFractionColumn] = new FractionState(numerator, denominator);

      const equal = fractionsCopy[FractionColumn.LEFT].value === fractionsCopy[FractionColumn.RIGHT].value;

      return {
        fractions: fractionsCopy,
        equal: equal
      };
  }

  /** Factory function to handle a denominator change from the specified fraction child object
  @param fractionColumn The fraction column that is setting up this event handler
  @return A function to update the state of this component in response to the denominator change
  */
  handleDenominatorChange(fractionColumn: FractionColumn) : (denominator: number) => void {
    return  (denominator: number) => {
        if(denominator > 0) {
          this.setState((state, props) => {
              // figure out the new fraction
              return this.buildState(this.state.fractions, state.fractions[fractionColumn].numerator, denominator, fractionColumn);
              });
          }
      };
  }

  /** Factory function to handle a numerator change from the specified fraction child object
  @param fractionIndex The fraction column that is setting up this event handler
  @return A function to update the state of this component in response to the denominator change
  */
  handleNumeratorChange(fractionColumn: FractionColumn) : (numerator: number) => void  {
    return (numerator: number) => {
        if(numerator >= 0) {
          this.setState((state, props) => {
            // figure out new fraction
            return this.buildState(this.state.fractions, numerator, state.fractions[fractionColumn].denominator, fractionColumn);
          });
        }
        
     };
  }

  /** Render the specified column
  @param fractionIndex The specified fraction column to render
  @param maxCharts The maximum number of charts in either column, so spacing is uniform
  @return The rendered output
  */
  private renderColumn(fractionColumn:FractionColumn, maxCharts: number) : JSX.Element {
       const fractionState = this.state.fractions[fractionColumn];
       const { numerator, denominator } = fractionState;

       return (
         <div className={styles.column}>
           <FractionInput  fractionState={fractionState} 
                   numeratorUpdated={this.handleNumeratorChange(fractionColumn)} 
                   denominatorUpdated={this.handleDenominatorChange(fractionColumn)} /> 

           <ChartContainer equal={this.state.equal} fractionState={fractionState} maxCharts={maxCharts} /> 
           <div className={styles.decimal}>{Math.round(100 * numerator/denominator)/100}</div>
         </div>
       );
  }

  render() : JSX.Element {
    const { fractions } = this.state;
    const maxCharts = Math.max(fractions[FractionColumn.LEFT].charts, fractions[FractionColumn.RIGHT].charts);
    
    return (
      <div className={styles.mainContainer}>
        { this.renderColumn(FractionColumn.LEFT, maxCharts) }
        { this.renderColumn(FractionColumn.RIGHT, maxCharts) }
      </div>
    );
  }

}
