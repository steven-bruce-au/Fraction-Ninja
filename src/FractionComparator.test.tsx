import { default as React, ReactElement } from 'react';
import util from 'util';
import { createRenderer } from 'react-test-renderer/shallow';
import * as renderer from 'react-test-renderer';

import { FractionState } from './FractionState';
import FractionComparator from './FractionComparator';
import FractionInput from './FractionInput';
import ChartContainer from './ChartContainer';

class EmptyParent implements renderer.TestRendererOptions {

   // override default implementation
   createNodeMock(element: ReactElement): any {
    const doc = document.implementation.createHTMLDocument();
    return { parentElement: doc.body };
   }
}

it('render FractionComparator', () => {

  const component = renderer.create(<FractionComparator/>, new EmptyParent());

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // console.log(util.inspect(tree, {showHidden: false, depth: null}));

});

function anonymous(numerator: number) : void {}


/** For some reason this doesn't work */
it.skip('render FractionComparator Shallow', () => {

const shallowRenderer = createRenderer();
shallowRenderer.render(<FractionComparator />);
const result = shallowRenderer.getRenderOutput();

console.log(util.inspect(result, {showHidden: false, depth: null}));
console.log(util.inspect(<div className="mainContainer">
  <div className="column">
       <FractionInput fractionState={new FractionState()} numeratorUpdated={ anonymous } denominatorUpdated={ anonymous }/>
       <ChartContainer equal={true} fractionState={ new FractionState() } maxCharts={1} />
       <div className="decimal">1</div>
  </div>
  <div className="column">
       <FractionInput fractionState={new FractionState()} numeratorUpdated={ anonymous } denominatorUpdated={ anonymous }/>
       <ChartContainer equal={true} fractionState={ new FractionState() } maxCharts={1} />
       <div className="decimal">1</div>
  </div>
</div>, {showHidden: false, depth: null}));




expect(result.type).toBe('div');
expect(result).toEqual(<div className="mainContainer">
  <div className="column">
       <FractionInput fractionState={ new FractionState() } numeratorUpdated={ anonymous } denominatorUpdated={ anonymous }/>
       <ChartContainer equal={true} fractionState={ new FractionState() } maxCharts={1} />
       <div className="decimal">1</div>
  </div>
  <div className="column">
       <FractionInput fractionState={ new FractionState() } numeratorUpdated={ anonymous } denominatorUpdated={ anonymous }/>
       <ChartContainer equal={true} fractionState={ new FractionState() } maxCharts={1} />
       <div className="decimal">1</div>
  </div>
</div>);

shallowRenderer.unmount()

});