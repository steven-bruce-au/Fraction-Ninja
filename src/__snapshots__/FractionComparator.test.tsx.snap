// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render FractionComparator 1`] = `
<div
  className="mainContainer"
>
  <div
    className="column"
  >
    <div
      className="fraction"
    >
      <div
        className="fractionRow"
      >
        <button
          className="button"
          onClick={[Function]}
        >
           - 
        </button>
        <div
          className="number underline numerator"
        >
           
          1
        </div>
        <button
          className="button"
          onClick={[Function]}
        >
           + 
        </button>
      </div>
      <div
        className="fractionRow"
      >
        <button
          className="button"
          onClick={[Function]}
        >
          -
        </button>
        <div
          className="number denominator"
        >
           
          1
        </div>
        <button
          className="button"
          onClick={[Function]}
        >
          +
        </button>
      </div>
    </div>
    <div
      className="chartContainer"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <div
        className="chart"
      >
        <div
          className="recharts-responsive-container"
          style={
            Object {
              "height": "100%",
              "maxHeight": undefined,
              "minHeight": undefined,
              "minWidth": undefined,
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "display": "none",
                "height": 0,
                "position": "absolute",
                "visibility": "hidden",
                "width": 0,
              }
            }
          />
        </div>
      </div>
    </div>
    <div
      className="decimal"
    >
      1
    </div>
  </div>
  <div
    className="column"
  >
    <div
      className="fraction"
    >
      <div
        className="fractionRow"
      >
        <button
          className="button"
          onClick={[Function]}
        >
           - 
        </button>
        <div
          className="number underline numerator"
        >
           
          1
        </div>
        <button
          className="button"
          onClick={[Function]}
        >
           + 
        </button>
      </div>
      <div
        className="fractionRow"
      >
        <button
          className="button"
          onClick={[Function]}
        >
          -
        </button>
        <div
          className="number denominator"
        >
           
          1
        </div>
        <button
          className="button"
          onClick={[Function]}
        >
          +
        </button>
      </div>
    </div>
    <div
      className="chartContainer"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <div
        className="chart"
      >
        <div
          className="recharts-responsive-container"
          style={
            Object {
              "height": "100%",
              "maxHeight": undefined,
              "minHeight": undefined,
              "minWidth": undefined,
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "display": "none",
                "height": 0,
                "position": "absolute",
                "visibility": "hidden",
                "width": 0,
              }
            }
          />
        </div>
      </div>
    </div>
    <div
      className="decimal"
    >
      1
    </div>
  </div>
</div>
`;
